cmaes/cma.py

```diff
@@ -278,9 +281,13 @@ def tell(self, solutions: List[Tuple[np.ndarray, float]]) -> None:
         ) * C_2.dot(y_w)
 
         norm_p_sigma = np.linalg.norm(self._p_sigma)
-        self._sigma *= np.exp(
-            (self._c_sigma / self._d_sigma) * (norm_p_sigma / self._chi_n - 1)
-        )
+        # To avoid overflow
+        _log_sigma = np.log(self._sigma + self._epsilon) + (self._c_sigma / self._d_sigma) * (
+                norm_p_sigma / self._chi_n - 1)
+        if _log_sigma > 10 ** 2.8:
+            self._sigma = np.exp(10 ** 2.8)
```

Comment:
Could you tell me the meaning of this magic number?

Reply:
Thanks for the question! I've just decided this magic number to avoid the overflow. It is more reasonable to set it as the maximum number of float. What do you think?

Reply:
Agree 👍  And you can refactor these changes like `self._sigma = np.exp(max(_log_sigma, sys.float_info.max))`.

Reply:
To avoid overflow, seemingly it needs to refactor like:

```
self._sigma = min(np.exp(_log_sigma), sys.float_info.max)
```

Reply:
Thanks for the suggestion! I addressed.

Reply:
I'm sorry. Due to my suggestion, the error will be raised again.

```
...
ValueError: Nan is detected: [nan nan nan nan nan nan]
/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/cmaes/cma.py:222: RuntimeWarning: invalid value encountered in sqrt
  D = np.sqrt(D2)
[E 2020-06-22 05:06:10,985] Nan is detected: [nan nan nan nan nan nan]
Traceback (most recent call last):
  File "/var/folders/9q/c1wp98sd4110kvnb89ycs7vj6xh7ks/T/.tmpLv0wtD", line 89, in <module>
    runner.run()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 172, in run
    while self._run_once():
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 186, in _run_once
    self._handle_ask_call(message)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 214, in _handle_ask_call
    trial = solver.ask(idg)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/optuna.py", line 86, in ask
    trial = self._create_new_trial()
(ALL) [00:01:37] [STUDIES   1787/1785 100%] [ETA  0s] canceled
(STUDY) [00:01:35] [STEPS   4033/8000  50%] [ETA  1m] "cma-es-median" "HPO-Bench-Protein"
(ALL) [00:01:37] [STUDIES   1788/1785 100%] [ETA  0s] canceled
...
```

Reply:
Hmm... I changed the revision to 58e67ec, 

```
$ git checkout 58e67ec
$ git status
HEAD detached at 58e67ec
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        fcnet_tabular_benchmarks/
        problems.json
        reproduce.sh
        results.json
        solvers.json
        studies.json

nothing added to commit but untracked files present (use "git add" to track)
$ python -m pip freeze | grep kurobako
kurobako==0.1.7
$ ./reproduce-error.sh
...
ValueError: Nan is detected: [nan nan nan nan nan nan]
/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/cmaes/cma.py:222: RuntimeWarning: invalid value encountered in sqrt
  D = np.sqrt(D2)
[E 2020-06-22 05:11:06,425] Nan is detected: [nan nan nan nan nan nan]
Traceback (most recent call last):
  File "/var/folders/9q/c1wp98sd4110kvnb89ycs7vj6xh7ks/T/.tmpMl1HVg", line 89, in <module>
    runner.run()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 172, in run
    while self._run_once():
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 186, in _run_once
    self._handle_ask_call(message)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 214, in _handle_ask_call
    trial = solver.ask(idg)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/optuna.py", line 86, in ask
    trial = self._create_new_trial()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/optuna.py", line 172, in _create_new_trial
    return optuna.trial.Trial(self._study, trial_id)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/optuna/trial/_trial.py", line 67, in __init__
    self._init_relative_params()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/optuna/trial/_trial.py", line 77, in _init_relative_params
    self.relative_params = self.study.sampler.sample_relative(
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/optuna/samplers/_cmaes.py", line 245, in sample_relative
    raise ValueError("Nan is detected: {}".format(params))
ValueError: Nan is detected: [nan nan nan nan nan nan]
/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/cmaes/cma.py:222: RuntimeWarning: invalid value encountered in sqrt
  D = np.sqrt(D2)
[E 2020-06-22 05:11:06,613] Nan is detected: [nan nan nan nan nan nan]
Traceback (most recent call last):
  File "/var/folders/9q/c1wp98sd4110kvnb89ycs7vj6xh7ks/T/.tmpMl1HVg", line 89, in <module>
    runner.run()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 172, in run
    while self._run_once():
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 186, in _run_once
    self._handle_ask_call(message)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 214, in _handle_ask_call
    trial = solver.ask(idg)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/optuna.py", line 86, in ask
(ALL) [00:01:43] [STUDIES   2117/2109 100%] [ETA  0s] canceled
(STUDY) [00:01:41] [STEPS   4097/8000  51%] [ETA  1m] "cma-es-median" "HPO-Bench-Protein"
/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/cmaes/cma.py:222: RuntimeWarning: invalid value encountered in sqrt
  D = np.sqrt(D2)
[E 2020-06-22 05:11:07,187] Nan is detected: [nan nan nan nan nan nan]
Traceback (most recent call last):
  File "/var/folders/9q/c1wp98sd4110kvnb89ycs7vj6xh7ks/T/.tmpMl1HVg", line 89, in <module>
    runner.run()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 172, in run
    while self._run_once():
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 186, in _run_once
    self._handle_ask_call(message)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/__init__.py", line 214, in _handle_ask_call
    trial = solver.ask(idg)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/optuna.py", line 86, in ask
    trial = self._create_new_trial()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/kurobako/solver/optuna.py", line 172, in _create_new_trial
    return optuna.trial.Trial(self._study, trial_id)
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/optuna/trial/_trial.py", line 67, in __init__
    self._init_relative_params()
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/optuna/trial/_trial.py", line 77, in _init_relative_params
    self.relative_params = self.study.sampler.sample_relative(
  File "/Users/a14737/src/github.com/CyberAgent/cmaes/venv/lib/python3.8/site-packages/optuna/samplers/_cmaes.py", line 245, in sample_relative
(ALL) [00:01:43] [STUDIES   2117/2109 100%] [ETA  0s] canceled

Error: InvalidInput (cause; EOF while parsing a value at line 1 column 0)
HISTORY:
  [0] at kurobako_core/src/epi/channel.rs:62 -- line=""
  [1] at kurobako_core/src/epi/solver/external_program.rs:164
  [2] at kurobako_core/src/epi/solver/embedded_script.rs:94
  [3] at kurobako_solvers/src/optuna.rs:276
  [4] at kurobako_core/src/solver.rs:176
  [5] at kurobako_core/src/solver.rs:176
  [6] at src/runner.rs:314
  [7] at src/runner.rs:271
  [8] at src/runner.rs:357
  [9] at src/runner.rs:136
  [10] at src/runner.rs:145
  [11] at src/main.rs:85


```

Reply:
Hmm... It seems the version of `cmaes` is wrong because the current `cmaes` in this PR does not contain `D = np.sqrt(D2)` but contain `D = np.sqrt(np.where(D2 < 0, 0, D2))`. Could you re-install `cmaes`?

Reply:
Oh, thanks. I'll re-try it.

Reply:
Thank you. The error seems not to be raised.

Now, I'm checking more simplified patch at #41.